This notebook documents my work for a data scientist position code test.  It has lightly been edited to include the questions and my answers (as per the multiple choice selection).  However, I have left the code as is so you can see my thought progress for each question.  I did the "CodeChallenge" questions within the broswer code editor and have not replicated them here.  My essay response was copy and pasted here right before I submitted it online. 

In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn

# Q1
You are given the data file 'data1.csv'.  The first line is the column header `[id, sla, overdue, x, y, type]`.

Find the mean value of the column `overdue`.

In [2]:
data1 = pd.read_csv('./AnalysisData/data1.csv')
data1.head()

id   sla  overdue           x            y  type
0           101000879701  10.0    -8.98  3105443.57  13832598.28   3.0
1  11682872-101001143202  14.0    -7.79  3047695.62  13826869.96   3.0
2           101001090815  10.0    -8.59  3078170.34  13820368.28   3.0
3  11594791-101000891316  15.0   -14.89  3041615.31  13820936.90   3.0
4           101001002655   4.0     4.57  3041051.24  13824386.21   3.0

In [10]:
data1.describe()

sla       overdue             x             y          type
count  25000.000000  25000.000000  2.500000e+04  2.500000e+04  25000.000000
mean      30.992280     -4.798978  3.109990e+06  1.383583e+07      2.945040
std       51.215171    102.036356  3.451559e+04  3.273077e+04      0.266725
min        1.000000   -501.040000  2.077445e+06  1.370344e+07      0.000000
25%        6.000000    -14.950000  3.085637e+06  1.381152e+07      3.000000
50%       14.000000     -6.100000  3.111065e+06  1.383432e+07      3.000000
75%       25.000000     -1.050000  3.133117e+06  1.385723e+07      3.000000
max      180.000000   1399.120000  3.207514e+06  1.396255e+07      5.000000

In [3]:
data1['overdue'].mean()

-4.798978000000018

## My Answer: -4.80

# Q2
For type 2 events, what is the median value of `sla`?

In [5]:
data1['sla'][data1['type'] == 2].median() 

14.0

## My Answer: 14.0

# Q3

What fraction of the events are `type` == 2?

In [6]:
data1['type'].value_counts()

3.0    23583
2.0     1321
5.0       41
0.0       31
1.0       22
4.0        2
Name: type, dtype: int64

In [15]:
(data1['type'][data1['type'] == 2].count()) / data1['type'].count()

0.05284

In [16]:
1321/25000 

0.05284

## My Answer: 0.05284

# Q4

You are given the data file 'data2.json'.  It contains a JSON array of objects with the following format:

- id: string
- id2: string
- key: string
- created: number, UNIX timestamp in milliseconds
- district: string

Find the 3 districts with the most entries.

In [35]:
data2 = pd.read_json('./AnalysisData/data2.json')
data2.head()

created district                     id            id2   key
0  1357571830000        C           101000879701  1231880010005  492U
1  1380367716000        F  11682872-101001143202  1181610010014  488Z
2  1376290781000        J           101001090815  0834050000002  530H
3  1358517113000        F  11594791-101000891316  1095590000010  528G
4  1368462905000        F           101001002655  1138180000022  528C

In [36]:
data2.describe()

created
count  2.500000e+04
mean   1.372960e+12
std    8.953443e+09
min    1.357029e+12
25%    1.365191e+12
50%    1.373450e+12
75%    1.380543e+12
max    1.388522e+12

In [80]:
data2['district'].value_counts()[:3]

C    3945
D    3338
B    2899
Name: district, dtype: int64

## My Answer: B, C, D

# Q5

Select the week that has the most entries.  Assume that the timestamps and dates are UTC.

In [38]:
data2['created'] = pd.to_datetime(data2['created'], unit='ms')

In [40]:
data2.dtypes

created     datetime64[ns]
district            object
id                  object
id2                 object
key                 object
dtype: object

In [41]:
data2.head()

created district                     id            id2   key
0 2013-01-07 15:17:10        C           101000879701  1231880010005  492U
1 2013-09-28 11:28:36        F  11682872-101001143202  1181610010014  488Z
2 2013-08-12 06:59:41        J           101001090815  0834050000002  530H
3 2013-01-18 13:51:53        F  11594791-101000891316  1095590000010  528G
4 2013-05-13 16:35:05        F           101001002655  1138180000022  528C

In [51]:
data2.shape

(25000, 6)

In [42]:
data2['created'].describe()

count                   25000
unique                  24979
top       2013-11-14 08:42:21
freq                        2
first     2013-01-01 08:26:29
last      2013-12-31 20:29:42
Name: created, dtype: object

In [44]:
data2['week'] = data2['created'].dt.week

In [48]:
data2['week'].value_counts()[:5]

33    623
39    582
38    544
32    541
42    531
Name: week, dtype: int64

In [50]:
data2[data2['week'] == 33][:10]

created district                     id            id2   key  \
2   2013-08-12 06:59:41        J           101001090815  0834050000002  530H   
36  2013-08-18 13:34:41        F         0-101001098519  1005100000013  529J   
67  2013-08-15 12:10:26        F  11665384-101001095518  1010990000001  529P   
71  2013-08-13 14:08:41        G  11664553-101001093189  0451400040271  491Z   
77  2013-08-12 06:11:22        D         0-101001090804  0170470110004  494W   
139 2013-08-16 13:48:10        A  11665889-101001097030  0780590080012  450Z   
162 2013-08-15 09:49:35        H           101001095292  0611890000012  493B   
210 2013-08-14 16:26:27        B  11665108-101001094720  0731800030578  454G   
279 2013-08-16 08:31:40        G  11665672-101001096395  1064180000056  489N   
298 2013-08-17 12:40:50        A  11666292-101001098061  0985480000015  450K   

     week  
2      33  
36     33  
67     33  
71     33  
77     33  
139    33  
162    33  
210    33  
279    33  
298    33

## My Answer: 12 - 18 Aug 2013

# Q6

This question asks you to use both datafiles from previous questions.

What day of the week has the lowest mean overdue value?

In [52]:
data_combine = pd.merge(data1, data2, how='outer',
                       left_on='id', right_on='id')

In [53]:
data_combine.head()

id   sla  overdue           x            y  type  \
0           101000879701  10.0    -8.98  3105443.57  13832598.28   3.0   
1  11682872-101001143202  14.0    -7.79  3047695.62  13826869.96   3.0   
2           101001090815  10.0    -8.59  3078170.34  13820368.28   3.0   
3  11594791-101000891316  15.0   -14.89  3041615.31  13820936.90   3.0   
4           101001002655   4.0     4.57  3041051.24  13824386.21   3.0   

              created district            id2   key  week  
0 2013-01-07 15:17:10        C  1231880010005  492U     2  
1 2013-09-28 11:28:36        F  1181610010014  488Z    39  
2 2013-08-12 06:59:41        J  0834050000002  530H    33  
3 2013-01-18 13:51:53        F  1095590000010  528G     3  
4 2013-05-13 16:35:05        F  1138180000022  528C    20

In [54]:
data_combine.shape

(25000, 11)

In [55]:
data_combine['dow'] = data_combine['created'].dt.dayofweek

In [57]:
data_combine.groupby('dow')['overdue'].mean()

dow
0    -4.413410
1    -3.808284
2    -3.825404
3    -1.377711
4    -6.304658
5    -9.960528
6   -13.976185
Name: overdue, dtype: float64

## My answer: Sunday

# Q7

Using both datafiles again, determine which district's mean location (`x` and `y` coordinates) is closest to the mean location of all data, ignoring data from district `Unknown`?

In [65]:
data_combine_dropUKN = data_combine.copy()
data_combine_dropUKN[data_combine_dropUKN['district'] == 'Unknown'] = np.NaN
data_combine_dropUKN.dropna()

(25000, 12)

In [67]:
data_combine_dropUKN['district'].value_counts()

C    3945
D    3338
B    2899
K    2777
H    2598
I    2340
A    1752
G    1578
F    1490
E    1254
J     990
Name: district, dtype: int64

In [70]:
d_coor = data_combine_dropUKN.groupby('district')[['x', 'y']].mean()
mean_x = data_combine_dropUKN['x'].mean()
mean_y = data_combine_dropUKN['y'].mean()

In [74]:
d_coor = d_coor.reset_index(drop=False)
d_coor

district             x             y
0         A  3.074090e+06  1.386535e+07
1         B  3.132169e+06  1.387285e+07
2         C  3.103778e+06  1.384182e+07
3         D  3.132290e+06  1.380794e+07
4         E  3.174524e+06  1.385730e+07
5         F  3.052256e+06  1.381715e+07
6         G  3.068135e+06  1.383798e+07
7         H  3.122190e+06  1.385954e+07
8         I  3.145095e+06  1.382659e+07
9         J  3.072780e+06  1.381760e+07
10        K  3.089845e+06  1.379485e+07

In [76]:
d_coor['dist'] = np.sqrt((d_coor['x'] - mean_x) ** 2 + (d_coor['y'] - mean_y) ** 2)

In [78]:
d_coor.sort_values('dist')

district             x             y          dist
2         C  3.103778e+06  1.384182e+07   8663.073735
7         H  3.122190e+06  1.385954e+07  26638.695141
3         D  3.132290e+06  1.380794e+07  35682.624983
8         I  3.145095e+06  1.382659e+07  36255.362938
9         J  3.072780e+06  1.381760e+07  41483.914571
6         G  3.068135e+06  1.383798e+07  41959.488050
1         B  3.132169e+06  1.387285e+07  43119.444928
10        K  3.089845e+06  1.379485e+07  45697.763803
0         A  3.074090e+06  1.386535e+07  46513.233446
5         F  3.052256e+06  1.381715e+07  60731.032206
4         E  3.174524e+06  1.385730e+07  67962.773610

## My Answer: C

# Free form essay question 

>Assume you are given multiple time series data sets that contain data covering the same time period from different sources (e.g. equipment sensor data, weather data, etc) and tasked with building a predictive model (e.g. forecasting one of the values).  Describe the process you would use to explore and prepare the data, and the process to select, train, and assess your model.  Do not assume that the data is uniformly sampled.


**My answer:**
## Exploration
I would do the initial EDA within a Jupyter Notebook, primarily using the pandas, numpy, matplotlib, seaborn packages.  First, I would look at data types, shapes, and completeness of each data set.  Then I would primarily be looking at the data sampling period for each data set, correlations present within each and across all data sets, look for potential cyclical trends within each data set.  Ideally, I would like to speak to a domain expert about where this data came from, how it was acquired, any "gotchas", etc.

## Preparation
Would drop any constant fields; employ imputation techniques for missing data and/or establish a rule for dropping rows/columns; encode meaningful categorical features and drop less meaningful ones (e.g. contain too many categories).  If user inputs are present, employ regex logic to combine spurious data and extract meaningful information.  Would need to resample the data to a common frequency, using various interpolation methods where necessary.  Additional features would probably be required or requested by the client/domain expert -- engineer those and additional features as necessary.  Finally, the numerical data would need to be scaled appropriately.  

## Select
The first model choice should always be something simple.  In terms of time-series data this should probably be an ARIMA-like model which aims to remove cyclical trends present within the data.  The second model choice would likely be an LSTM-like model to gain the benefits of a neural-net model.

## Train
Given the prior EDA, select representative time periods to use for train/validation/test splits.  Grid search the hyperparameter space appropriately.  Evaluate the trained models using the validation set and the established model metric.  Perhaps training time or model size might be a consideration too.

## Evaluation
Work with the client/domain expert to establish a meaningful model metric based on the target variable.  Having client buy-in on model evaluation is critical for ultimate acceptance and usage of the model.  This metric will naturally differ for regression and classification models.  It could either be an error-term related value or an accuracy related value.


